modules/images/webp-uploads/load.php

```diff
@@ -137,7 +140,10 @@ function webp_uploads_create_sources_property( array $metadata, $attachment_id )
 			$properties['sources'] = array();
 		}
 
-		if ( empty( $properties['sources'][ $current_mime ] ) ) {
+		if (
+			empty( $properties['sources'][ $current_mime ] ) &&
+			in_array( $current_mime, $valid_mime_transforms[ $mime_type ], true )
```

Comment:
Why is this needed? If the size already has a file of `$current_mime`, whatever the MIME type is WordPress would only have generated it if it should. If without this condition we get a problem, then I think the bug is somewhere else.

Reply:
If I remember correctly, the problem was that we kept generating source information for the `image/jpeg` mime type despite the fact that we were no jpeg images after converting it to webp. This fix ensures that we don't add a mime type to the sources list if it doesn't exists in the transformations list.

Reply:
@eugene-manuilov

> we kept generating source information for the `image/jpeg` mime type despite the fact that we were no jpeg images after converting it to webp.

This is one of the problems I was trying to convey further above. We need to fix _that_, rather than add this condition here I think.

Reply:
I tested this locally, and I still think the extra condition here should be reverted. With the configuration of a single MIME type, `$current_mime` will be set to `image/webp`, as that is the MIME type of any existing image size WP core generates in that scenario. All this condition does is set the `sources` information for that existing WebP image. Checking whether it is in the list of allowed transform MIME types is unnecessary - mostly since that file already exists. If the file already exists, we don't need to check whether our filters allow for it. It's already generated, so we can just populate the information for it.

The MIME transforms filter is critical only to decide whether to transform an image of a certain format to a certain format, but for an image that already exists we can just add its information. In fact, we need to remove a similar condition around line 58 as well, since otherwise it runs into the problem that for the original JPEG no `sources` value is added, just because we don't _want_ to generate it.

Reply:
@eugene-manuilov I fixed this in https://github.com/WordPress/performance/pull/527/commits/670b63f9ae5679af1f4ffd7556ad7fdd04477f75. There are currently unit test failures, but those are only happening because _WordPress_ still keeps the full JPEG image. Simply not putting it into `sources` should not be a fix for those tests, since that means the image is still there.

What I think we _want_ to do in the case that only WebP should be generated is actually override the original file with our generated "full" size WebP. We can use `_wp_image_meta_replace_original()` for that. And in that case, we can then also remove that original image from `sources`.

Reply:
@eugene-manuilov I found the actual problem that was causing the test to fail, and fixed it in https://github.com/WordPress/performance/pull/527/commits/8825df789b16790dfa4e3d2fe88717fed2e23125 (see code comment on what was causing the problem).

The extra condition here is indeed not necessary and was only avoiding the failure rather than fixing the problem which was more in the test itself.